backends/gstreamer/src/audio_sink.rs

```diff
@@ -23,15 +23,28 @@ pub struct GStreamerAudioSink {
     sample_offset: Cell<u64>,
 }
 
+#[derive(Debug)]
+pub enum BackendError {
+    Gstreamer(gst::Error),
+    Flow(gst::FlowError),
+    ElementCreationFailed(&'static str),
+    DowncastingFailed,
+    AudioInfoFailed,
+    PipelineFailed,
+    StateChangeFailed,
+}
+
+
 impl GStreamerAudioSink {
-    pub fn new() -> Result<Self, ()> {
+    pub fn new() -> Result<Self, BackendError> {
         if let Some(category) = gst::DebugCategory::get("openslessink") {
             category.set_threshold(gst::DebugLevel::Trace);
         }
-        gst::init().map_err(|_| ())?;
+        gst::init().map_err(BackendError::Gstreamer)?;
 
-        let appsrc = gst::ElementFactory::make("appsrc", None).ok_or(())?;
-        let appsrc = appsrc.downcast::<AppSrc>().map_err(|_| ())?;
+        let appsrc = gst::ElementFactory::make("appsrc", None)
+                .ok_or(BackendError::ElementCreationFailed("appsrc"))?;
+        let appsrc = appsrc.downcast::<AppSrc>().map_err(|_| BackendError::DowncastingFailed)?;
         Ok(Self {
             pipeline: gst::Pipeline::new(None),
             appsrc: Arc::new(appsrc),
```

Comment:
Why is the appsrc not put into the pipeline here already?

Also generally it seems like for *all* pipelines you're not watching their `gst::Bus` for messages, especially not for error messages. It would be good to add some code for that so that you can handle errors (and later possibly other things) asynchronously while the pipeline is running

Reply:
> Also generally it seems like for all pipelines you're not watching their gst::Bus for messages, especially not for error messages. It would be good to add some code for that so that you can handle errors (and later possibly other things) asynchronously while the pipeline is running

Yeah, we'll need that for #125